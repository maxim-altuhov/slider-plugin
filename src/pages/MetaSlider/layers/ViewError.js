class ViewError {
  constructor() {
    this.presenter = null;
  }

  registerWith(presenter) {
    this.presenter = presenter;
  }

  getProp(prop) {
    return this.presenter.getProp(prop);
  }

  setProp(prop, value) {
    this.presenter.setProp(prop, value);
  }

  init() {
    this.$selector = this.getProp('$initSelector');
    this.$elemSlider = this.getProp('$elemSlider');
  }

  renderError(message) {
    if (this.getProp('showError')) {
      const HTMLBlockError = `<div class="error-info js-error-info"><p class="error-info__text">${message}</p><button type="button" class="error-info__close js-error-info__close">X</button></div>`;
      this.$elemSlider.after(HTMLBlockError);

      this.getInfoAboutError();
      this.setEventErrorClose();
      this.setProp('showError', false);
    }
  }

  getInfoAboutError() {
    this.$elemErrorInfo = this.$selector.find('.js-error-info');
    this.$btnErrorClose = this.$selector.find('.js-error-info__close');
  }

  handleRemoveErrorWindow() {
    this.$elemErrorInfo.remove();
    this.setProp('showError', true);
  }

  setEventErrorClose() {
    if (this.$btnErrorClose) {
      this.$btnErrorClose.on('click.btnErrorClose', this.handleRemoveErrorWindow.bind(this));
    }
  }
}

export default ViewError;
