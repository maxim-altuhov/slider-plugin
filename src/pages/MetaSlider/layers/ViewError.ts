class ViewError {
  $selector: undefined | JQuery;
  $elemErrorInfo: null | JQuery;
  $elemErrorText: null | JQuery;
  $btnErrorClose: null | JQuery;

  // Метод для рендеринга ошибки
  renderError(message: string, options: IPluginOptions) {
    const { $selector, $elemSlider, showError } = options;

    if (!this.$selector) this.$selector = $selector;

    if (showError && !this.$elemErrorInfo) {
      const HTMLBlockError = `<div class="error-info js-error-info">
      <p class="error-info__text js-error-info__text">${message}</p>
      <button type="button" class="error-info__close js-error-info__close">X</button>
      </div>`;
      $elemSlider.after(HTMLBlockError);

      this._getInfoAboutError();
      this._setEventErrorClose();
    } else if (showError && this.$elemErrorInfo) {
      this.$elemErrorText.text(message);
    }
  }

  // Собираем селекторы
  private _getInfoAboutError() {
    this.$elemErrorInfo = this.$selector.find('.js-error-info');
    this.$elemErrorText = this.$selector.find('.js-error-info__text');
    this.$btnErrorClose = this.$selector.find('.js-error-info__close');
  }

  // Устанавливаем обработчик на кнопку закрытия окна с ошибкой
  private _setEventErrorClose() {
    if (this.$btnErrorClose) {
      this.$btnErrorClose.on('click.btnErrorClose', this._handleRemoveErrorWindow.bind(this));
    }
  }

  // Удаляем сообщение с ошибкой
  private _handleRemoveErrorWindow() {
    this.$btnErrorClose.off('click.btnErrorClose');
    this.$elemErrorInfo.remove();
    this.$elemErrorInfo = null;
    this.$elemErrorText = null;
    this.$btnErrorClose = null;
  }
}

export default ViewError;
