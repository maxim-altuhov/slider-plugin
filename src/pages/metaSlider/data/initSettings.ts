const initSettings: IPluginOptions = {
  key: '',
  $selector: $(),
  $elemSlider: $(),
  $sliderProgress: $(),
  $elemMarkers: $(),
  $elemScale: $(),
  $elemThumbs: $(),
  mainColor: '#6d6dff',
  secondColor: '#e4e4e4',
  colorMarker: '',
  colorThumb: '',
  colorTextForMarker: '#ffffff',
  colorBorderForMarker: '#ffffff',
  colorBorderForThumb: '#ffffff',
  colorForScale: '#000000',
  initFormatted: true,
  initAutoMargins: true,
  initScaleAdjustment: true,
  calcNumberOfDecimalPlaces: true,
  showError: false,
  showScale: true,
  showMarkers: true,
  showBackground: true,
  isRange: true,
  isVertical: false,
  initAutoScaleCreation: true,
  checkingStepSizeForScale: false,
  step: 1,
  stepAsPercent: 0,
  minValue: 0,
  maxValue: 0,
  stepSizeForScale: null,
  numberOfDecimalPlaces: 0,
  preFix: '',
  postFix: '',
  customValues: [],
  initValueFirst: null,
  initValueSecond: null,
  initValuesArray: [],
  textValueFirst: '',
  textValueSecond: '',
  textValuesArray: [],
  valuesAsPercentageArray: [],
};

export default initSettings;
