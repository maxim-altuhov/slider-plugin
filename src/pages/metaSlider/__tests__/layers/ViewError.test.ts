/* eslint-disable @typescript-eslint/dot-notation */
import ViewError from '../../layers/ViewError';
import initSettings from '../../data/initSettings';

const initHTMLBlock = '<div class="fake-selector"><div class="fake-slider"></div></div>';
document.body.innerHTML = initHTMLBlock;

describe('Checking the "ViewError" layer, before first initialization "renderError" method', () => {
  const notInitViewError = new ViewError();

  test('State before first initialization', () => {
    expect(document.body.innerHTML).toMatchSnapshot();
    expect(notInitViewError['_$selector']).toBeUndefined();
    expect(notInitViewError['_$elemErrorInfo']).toBeUndefined();
    expect(notInitViewError['_$elemErrorText']).toBeUndefined();
    expect(notInitViewError['_$elemErrorCloseBtn']).toBeUndefined();
  });
});

describe('Checking the "ViewError" layer', () => {
  const classViewError = new ViewError();
  const TEXT_MSG = 'TEXT MESSAGE';
  const NEW_TEXT_MSG = 'NEW TEXT';

  const defaultSettings = {
    $selector: $('.fake-selector'),
    $elemSlider: $('.fake-slider'),
    showError: true,
  };

  const testSettings: IPluginOptions = $.extend({}, initSettings, defaultSettings);
  let $elemErrorInfo: null | JQuery<HTMLElement> | undefined;
  let $elemErrorText: null | JQuery<HTMLElement> | undefined;
  let $elemErrorCloseBtn: null | JQuery<HTMLElement> | undefined;

  beforeEach(() => {
    classViewError.renderError(TEXT_MSG, testSettings);

    $elemErrorInfo = classViewError['_$elemErrorInfo'];
    $elemErrorText = classViewError['_$elemErrorText'];
    $elemErrorCloseBtn = classViewError['_$elemErrorCloseBtn'];
  });

  test('Checking the "renderError" method => Creating block with error', () => {
    const documentPage = document.body.innerHTML;

    expect(classViewError['_$selector']).toHaveLength(1);
    expect(documentPage).not.toBe(initHTMLBlock);
    expect(documentPage).toMatch(/js-error-info/);
    expect(documentPage).toMatch(/js-error-info__text/);
    expect(documentPage).toMatch(/js-error-info__close-btn/);
  });

  test('Checking the "_getErrorBlockSelectors" method => Data about selectors is collected', () => {
    expect($elemErrorInfo).toHaveLength(1);
    expect($elemErrorText).toHaveLength(1);
    expect($elemErrorCloseBtn).toHaveLength(1);
  });

  test('Checking the "renderError" method => The error text is output', () => {
    expect($elemErrorText?.text()).toBe(TEXT_MSG);
  });

  test('Checking the "renderError" method => If block already created, just changed the error message', () => {
    expect($elemErrorText).toHaveLength(1);
    expect($elemErrorText?.text()).toBe(TEXT_MSG);

    classViewError.renderError(NEW_TEXT_MSG, testSettings);

    expect($elemErrorText?.text()).toBe(NEW_TEXT_MSG);
  });

  test('Checking the "_setEventErrorClose" method, event "click" => init _handleCloseBtnClick => When we click on the button to close the error window, it is deleted and all selectors are reset', () => {
    expect(classViewError['_$elemErrorInfo']).not.toBeNull();
    expect(classViewError['_$elemErrorText']).not.toBeNull();
    expect(classViewError['_$elemErrorCloseBtn']).not.toBeNull();
    expect(document.body.innerHTML).not.toBe(initHTMLBlock);

    classViewError['_$elemErrorCloseBtn']?.trigger('click');

    expect(classViewError['_$elemErrorInfo']).toBeNull();
    expect(classViewError['_$elemErrorText']).toBeNull();
    expect(classViewError['_$elemErrorCloseBtn']).toBeNull();
    expect(document.body.innerHTML).toBe(initHTMLBlock);
  });
});
