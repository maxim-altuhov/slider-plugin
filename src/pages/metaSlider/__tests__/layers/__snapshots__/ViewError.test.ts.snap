// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checking the "ViewError" layer, before first initialization "renderError" method State before first initialization 1`] = `"<div class=\\"fake-selector\\"><div class=\\"fake-slider\\"></div></div>"`;
